# It is a regression problem 

## We are doing the machine learning on house price data set 

<table align="center" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b> Regression modeling on House price data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

# Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

In [1]:
# pandas is utilized to format the data sturctures it means it can convert any data sturcture into pandas dataframe
# import 'Pandas'
import pandas as pd

# to find stastical summaries we use numpy
# import 'Numpy'
import numpy as np

# import subpackage of Matplotlib
import matplotlib.pyplot as plt

### Load the Dataset 

In [2]:
# Load the date and name it df_house_price
df_houseprice = pd.read_csv('House_Price.csv')

## Data Overview

In [3]:
# name the columns
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `price`: price of the house.
 - `resid_area`: Residential area where my house is located
 - `air_qual`: Air Quality

In [4]:
# Basic info 0f the dateset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Intrepretation**

 - We have 13 numerical columns & 3 categorical column
 - We have NULL value in the date
 - We have total  18 columns in the data

In [5]:
# Shape of the data
df_houseprice.shape

(506, 16)

**Intrepretation**

 - We have total 16 columns
 - We have 506 observations

In [6]:
# Summrise the basic statics of the data
df_houseprice.describe()

price  resid_area    air_qual    room_num         age    distance  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    22.528854   41.136779    0.554695    6.284634   68.574901    3.971996   
std      9.182176    6.860353    0.115878    0.702617   28.148861    2.108532   
min      5.000000   30.460000    0.385000    3.561000    2.900000    1.130000   
25%     17.025000   35.190000    0.449000    5.885500   45.025000    2.270000   
50%     21.200000   39.690000    0.538000    6.208500   77.500000    3.385000   
75%     25.000000   48.100000    0.624000    6.623500   94.075000    5.367500   
max     50.000000   57.740000    0.871000    8.780000  100.000000   12.320000   

         teachers   poor_prop  n_hos_beds  n_hot_rooms    rainfall  \
count  506.000000  506.000000  498.000000   506.000000  506.000000   
mean    21.544466   12.653063    7.899767    13.041605   39.181818   
std      2.164946    7.141062    1.476683     5.238957   12.513697   
min     18.000000    1.730000    5.268000    10.057600    3.000000   
25%     19.800000    6.950000    6.634500    11.189800   28.000000   
50%     20.950000   11.360000    7.999000    12.720000   39.000000   
75%     22.600000   16.955000    9.088000    14.170800   50.000000   
max     27.400000   37.970000   10.876000   101.120000   60.000000   

            parks        Sold  
count  506.000000  506.000000  
mean     0.054454    0.454545  
std      0.010632    0.498422  
min      0.033292    0.000000  
25%      0.046464    0.000000  
50%      0.053507    0.000000  
75%      0.061397    1.000000  
max      0.086711    1.000000

**Intrepretation**

 - We have outliers in the data, most of the outliers are in age column

# Data preprocessing

## a. Data cleaning

In [7]:
# show the first five observations
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [8]:
# Rename the columns 
# step1: show the columns Names
df_houseprice.columns


Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [9]:
# step2: Allocate New column names to Old column names
# Make the Dictionary('Old column anme':'New column name')
d1={
    'price':'Price',
    'resid_area':'Residential_Area',
    'air_qual':'Air_Quality',
    'room_num':'Room_Numbers', 
    'age':'Age', 
    'distance':'Distance',
    'teachers':'Teachers', 
    'poor_prop':'Proportion_Of_Population', 
    'airport':'Airport', 
    'n_hos_beds':'Number_Of_Hospital_Beds', 
    'n_hot_rooms':'Number_Of_Hotel_Rooms', 
    'waterbody':'Waterbody', 
    'rainfall':'Rainfall', 
    'bus_ter':'Bus_Terminal', 
    'parks':'Parks', 
    'Sold':'Sold',
   }

In [10]:
df_houseprice_1 = df_houseprice.rename(columns=d1)

### b. Null Value Handelling

In [11]:
# find the valid null values
df_houseprice_1.isnull().sum()

Price                         0
Residential_Area              0
Air_Quality                   0
Room_Numbers                  0
Age                           0
Distance                      0
Teachers                      0
Proportion_Of_Population      0
Airport                       0
Number_Of_Hospital_Beds       8
Number_Of_Hotel_Rooms         0
Waterbody                   155
Rainfall                      0
Bus_Terminal                  0
Parks                         0
Sold                          0
dtype: int64

In [12]:
# %missing values
round(df_houseprice_1.isnull().sum())/len(df_houseprice_1)*100

Price                        0.000000
Residential_Area             0.000000
Air_Quality                  0.000000
Room_Numbers                 0.000000
Age                          0.000000
Distance                     0.000000
Teachers                     0.000000
Proportion_Of_Population     0.000000
Airport                      0.000000
Number_Of_Hospital_Beds      1.581028
Number_Of_Hotel_Rooms        0.000000
Waterbody                   30.632411
Rainfall                     0.000000
Bus_Terminal                 0.000000
Parks                        0.000000
Sold                         0.000000
dtype: float64

In [13]:
df_houseprice_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Price                     506 non-null    float64
 1   Residential_Area          506 non-null    float64
 2   Air_Quality               506 non-null    float64
 3   Room_Numbers              506 non-null    float64
 4   Age                       506 non-null    float64
 5   Distance                  506 non-null    float64
 6   Teachers                  506 non-null    float64
 7   Proportion_Of_Population  506 non-null    float64
 8   Airport                   506 non-null    object 
 9   Number_Of_Hospital_Beds   498 non-null    float64
 10  Number_Of_Hotel_Rooms     506 non-null    float64
 11  Waterbody                 351 non-null    object 
 12  Rainfall                  506 non-null    int64  
 13  Bus_Terminal              506 non-null    object 
 14  Parks     

In [14]:
# This is a categorical column so we can replace the null with mod

In [15]:
df_houseprice_1.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [16]:
# Repalce the null is that column
df_houseprice_1.Waterbody.fillna('River', inplace=True)

C:\Users\anura\AppData\Local\Temp\ipykernel_19404\299150474.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_houseprice_1.Waterbody.fillna('River', inplace=True)


In [17]:
# %missing values
round(df_houseprice_1.isnull().sum())/len(df_houseprice_1)*100

Price                       0.000000
Residential_Area            0.000000
Air_Quality                 0.000000
Room_Numbers                0.000000
Age                         0.000000
Distance                    0.000000
Teachers                    0.000000
Proportion_Of_Population    0.000000
Airport                     0.000000
Number_Of_Hospital_Beds     1.581028
Number_Of_Hotel_Rooms       0.000000
Waterbody                   0.000000
Rainfall                    0.000000
Bus_Terminal                0.000000
Parks                       0.000000
Sold                        0.000000
dtype: float64

In [18]:
# In Number_Of_Hospital_Beds contain less than 10% values so we can drop null values
df_houseprice_1.dropna(inplace=True)

In [19]:
null_percentage = (df_houseprice_1.isnull().sum() / len(df_houseprice_1)) * 100
null_percentage

Price                       0.0
Residential_Area            0.0
Air_Quality                 0.0
Room_Numbers                0.0
Age                         0.0
Distance                    0.0
Teachers                    0.0
Proportion_Of_Population    0.0
Airport                     0.0
Number_Of_Hospital_Beds     0.0
Number_Of_Hotel_Rooms       0.0
Waterbody                   0.0
Rainfall                    0.0
Bus_Terminal                0.0
Parks                       0.0
Sold                        0.0
dtype: float64

## Preapare data for machine learning

In [20]:
# Seperate the terget variable and independent variable
target = df_houseprice_1['Sold']

In [21]:
# Independant variable
df_Independent = df_houseprice_1.drop('Sold', axis = 1)